pkg/pillar/cmd/zfsmanager/handlediskconfig.go

```diff
@@ -0,0 +1,211 @@
+	for _, diskCfg := range disks.Disks {
+		diskName := filepath.Base(diskCfg.Disk.Name)
+		if diskCfg.OldDisk != nil {
+			oldDiskName := filepath.Base(diskCfg.OldDisk.Name)
+			oldDevName, oldDevStat := getVdevNameStatus(vdevTree, oldDiskName)
+			if oldDevName != "" && diskCfg.Disk.Name != "" {
+				// if we found device oldDevName, replace it
+				log.Functionf("replacing %s with %s, old stat %s", oldDevName, diskCfg.Disk.Name, oldDevStat.State.String())
+				if stdout, err := zfs.ReplaceVDev(log, vault.DefaultZpool, oldDevName, diskCfg.Disk.Name); err != nil {
+					log.Errorf("cannot replace %s with %s: %s %s", oldDevName, diskCfg.Disk.Name, stdout, err)
+					continue
+				}
+			}
+		}
+		devName, devStat := getVdevNameStatus(vdevTree, diskName)
+		// found in pool
+		if devName != "" {
+			log.Functionf("zpool config disk %s, op %d, stat %s", devName, diskCfg.Config, devStat.State.String())
+			switch diskCfg.Config {
+			case types.EdgeNodeDiskConfigTypeZfsOnline:
+				switch zfs.GetZfsDeviceStatusFromStr(devStat.State.String()) {
+				case info.StorageStatus_STORAGE_STATUS_OFFLINE:
+					if err := pool.Online(true, devName); err != nil {
+						log.Errorf("cannot bring %s online: %s", devName, err)
+					}
+				case info.StorageStatus_STORAGE_STATUS_ONLINE:
+					continue
+				default:
+					log.Errorf("unexpected state of disk %s (%s) to make online", devName, devStat.State.String())
+					continue
+				}
+			case types.EdgeNodeDiskConfigTypeZfsOffline:
+				switch zfs.GetZfsDeviceStatusFromStr(devStat.State.String()) {
+				case info.StorageStatus_STORAGE_STATUS_ONLINE:
+					if err := pool.Offline(true, devName); err != nil {
+						log.Errorf("cannot bring %s offline: %s", devName, err)
+					}
+				case info.StorageStatus_STORAGE_STATUS_OFFLINE:
+					continue
+				default:
+					log.Errorf("unexpected state of disk %s (%s) to make offline", devName, devStat.State.String())
+					continue
+				}
+			case types.EdgeNodeDiskConfigTypeUnused:
+				if stdout, err := zfs.RemoveVDev(log, vault.DefaultZpool, devName); err != nil {
+					log.Errorf("cannot remove %s: %s %s", devName, stdout, err)
+				}
+			}
+		}
+	}
+	for _, el := range disks.Children {
+		disksStateProcessing(el, pool)
+	}
+}
+
+//disksLayoutProcessing iterate over disks and adjust pool layout accordingly to the config
+func disksLayoutProcessing(disks types.EdgeNodeDisks, pool libzfs.Pool) {
+	vdevTree, err := pool.VDevTree()
+	if err != nil {
+		log.Errorf("cannot get vdev tree: %s", err)
+		return
+	}
+	// TBD: other array types
+	if disks.ArrayType == types.EdgeNodeDiskArrayTypeRAID0 {
+		for _, el := range disks.Children {
+			if el.ArrayType == types.EdgeNodeDiskArrayTypeRAID1 {
+				diskName := ""
+				diskDescriptionName := ""
+				for _, dsk := range el.Disks {
+					if currentDiskName, _ := getVdevNameStatus(vdevTree, dsk.Disk.Name); currentDiskName != "" {
+						diskName = currentDiskName
+						diskDescriptionName = dsk.Disk.Name
+						break
+					}
+				}
```

Comment:
Would it make sense to define this as some lookupDiskName function?

Reply:
I simplified the logic a bit, actually it is not a lookup, but just the check that any of disks from the slice exists inside the tree.